Mathlib/Analysis/Convex/Function.lean

```diff
@@ -0,0 +1,1249 @@
+
+/-- A linear map is convex. -/
+theorem LinearMap.convexOn (f : E →ₗ[𝕜] β) {s : Set E} (hs : Convex 𝕜 s) : ConvexOn 𝕜 s f :=
+  ⟨hs, fun _ _ _ _ _ _ _ _ _ => by rw [f.map_add, f.map_smul, f.map_smul]⟩
+#align linear_map.convex_on LinearMap.convexOn
+
+/-- A linear map is concave. -/
+theorem LinearMap.concaveOn (f : E →ₗ[𝕜] β) {s : Set E} (hs : Convex 𝕜 s) : ConcaveOn 𝕜 s f :=
+  ⟨hs, fun _ _ _ _ _ _ _ _ _ => by rw [f.map_add, f.map_smul, f.map_smul]⟩
+#align linear_map.concave_on LinearMap.concaveOn
+
+theorem StrictConvexOn.convexOn {s : Set E} {f : E → β} (hf : StrictConvexOn 𝕜 s f) :
+    ConvexOn 𝕜 s f :=
+  convexOn_iff_pairwise_pos.mpr
+    ⟨hf.1, fun _ hx _ hy hxy _ _ ha hb hab => (hf.2 hx hy hxy ha hb hab).le⟩
+#align strict_convex_on.convex_on StrictConvexOn.convexOn
+
+set_option synthInstance.etaExperiment true in -- lean4#2074
+theorem StrictConcaveOn.concaveOn {s : Set E} {f : E → β} (hf : StrictConcaveOn 𝕜 s f) :
+    ConcaveOn 𝕜 s f :=
+  hf.dual.convexOn
+#align strict_concave_on.concave_on StrictConcaveOn.concaveOn
+
+section OrderedSMul
+
+variable [OrderedSMul 𝕜 β] {s : Set E} {f : E → β}
+
+theorem StrictConvexOn.convex_lt (hf : StrictConvexOn 𝕜 s f) (r : β) :
+    Convex 𝕜 ({ x ∈ s | f x < r }) :=
+  convex_iff_pairwise_pos.2 fun x hx y hy hxy a b ha hb hab =>
+    ⟨hf.1 hx.1 hy.1 ha.le hb.le hab,
+      calc
+        f (a • x + b • y) < a • f x + b • f y := hf.2 hx.1 hy.1 hxy ha hb hab
+        _ ≤ a • r + b • r :=
+          (add_le_add (smul_lt_smul_of_pos hx.2 ha).le (smul_lt_smul_of_pos hy.2 hb).le)
+        _ = r := Convex.combo_self hab r
+        ⟩
+#align strict_convex_on.convex_lt StrictConvexOn.convex_lt
+
+set_option synthInstance.etaExperiment true in -- lean4#2074
+theorem StrictConcaveOn.convex_gt (hf : StrictConcaveOn 𝕜 s f) (r : β) :
+    Convex 𝕜 ({ x ∈ s | r < f x }) :=
+  hf.dual.convex_lt r
+#align strict_concave_on.convex_gt StrictConcaveOn.convex_gt
+
+end OrderedSMul
+
+section LinearOrder
+
+variable [LinearOrder E] {s : Set E} {f : E → β}
+
+/-- For a function on a convex set in a linearly ordered space (where the order and the algebraic
+structures aren't necessarily compatible), in order to prove that it is convex, it suffices to
+verify the inequality `f (a • x + b • y) ≤ a • f x + b • f y` only for `x < y` and positive `a`,
+`b`. The main use case is `E = 𝕜` however one can apply it, e.g., to `𝕜^n` with lexicographic order.
+-/
+theorem LinearOrder.convexOn_of_lt (hs : Convex 𝕜 s)
+    (hf : ∀ ⦃x⦄, x ∈ s → ∀ ⦃y⦄, y ∈ s → x < y → ∀ ⦃a b : 𝕜⦄, 0 < a → 0 < b → a + b = 1 →
+      f (a • x + b • y) ≤ a • f x + b • f y) :
+    ConvexOn 𝕜 s f := by
+  refine' convexOn_iff_pairwise_pos.2 ⟨hs, fun x hx y hy hxy a b ha hb hab => _⟩
+  wlog h : x < y
+  · rw [add_comm (a • x), add_comm (a • f x)]
+    rw [add_comm] at hab
+    -- `wlog` gives a bad term, which has picked up all the stray variables.
+    -- See https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/wlog.20.2316495
```

Comment:
I think using `wlog generalizing` would be better than listing every variable here!